layout/src/main/java/com/vlkan/log4j2/logstash/layout/resolver/TemplateResolvers.java

```diff
@@ -158,55 +158,66 @@ public void resolve(V value, JsonGenerator jsonGenerator) throws IOException {
             }
         }
 
-        // Check if substitution is possible.
-        @SuppressWarnings("StringEquality") final boolean substitutionSupported =
-                fieldValue != context.getSubstitutor().replace(null, fieldValue);
+        // This is the fallback template resolver that delegates every other substitution
+        // to Log4j. This will be the case for every template value that does not use ${json:xxx}
+        // This is especially useful because it serves as a mechanism to resolve values at runtime
+        // when this library does not offer specific resolvers for the usecase.
 
-        if (substitutionSupported) {
-            if (LogEvent.class.isAssignableFrom(context.getContextClass())) {
-                // Use Log4j substitutor with LogEvent.
-                return new TemplateResolver<V>() {
+        // First try to resolve the value at initialisation phase. This will avoid having to resolve
+        // the same value multiple times at runtime.
+        // If the string remains the same, then Log4j was not able to replace the provided value with a
+        // resolved value.
+        final String staticallyResolvedValue = context.getSubstitutor().replace(null, fieldValue);
 
-                    @Override
-                    public void resolve(V value, JsonGenerator jsonGenerator) throws IOException {
+        @SuppressWarnings("StringEquality")
+        final boolean unresolved = (fieldValue == staticallyResolvedValue);
```

Comment:
not so sure about this string equality so I left it as it was before. why would we not use the String's `equals` method instead of `==`?

Reply:
An input `foo` would have been replaced but still had produced the same output `foo`. IIRC, substitutor returns the input to let the caller know about the substitution. Hence, we perform an object identity check, rather than content equality. That said, I will double check this -- I am still reviewing the changes whenever I find time.